### ISSUE TITLE
Holy water doesn't work

### ISSUE THREAD

**Issue by rmj555** [wesnoth/wesnoth#5385]
1.15.8
Windows 10
Sourceforge

Holy water did make a melee weapon arcane in either HttT or TRoW

**Comment by rmj555** [comment]
Paladins don't have arcane weapons.

**Comment by newfrenchy83** [comment]
@rmj555 paladins have already a arcane sword, and fore me holy water work without 'fix' what i have added in my PR, and the unit who are already an arcane attack are exluded by the filter in macros/items.cfg line 271

**Comment by CelticMinstrel** [comment]
I'm not sure what point either of you are trying to say about paladins, but let me state for the record what _should_ happen if a paladin tries to take holy water: it should say the paladin cannot use it.

**Comment by sevu** [comment]
That _should_ already be the case.

**Comment by AI0867** [comment]
A paladin also has a lance though. Arguably, holy water should be usable to make that weapon arcane too.

**Comment by rmj555** [comment]
The problem, it seems, is not that a weapon does not become arcane, but that the weapon is not described as arcane.  I had two royal guards; one hit skeletons at 12-4, the other at 5-4.

**Comment by soliton-** [comment]
Described where, @rmj555?

If you're talking about the help pages then those describe the unit type.

**Comment by rmj555** [comment]
When attacking you get a choice of weapons. The melee weapon is not described as arcane.

**Comment by soliton-** [comment]
Can anyone reproduce this or can you provide a save that shows the issue?

**Comment by doofus-01** [comment]
Weapons aren't described as 'arcane' or 'pierce' in the attack dialog.  Specials get a string, but 'arcane' isn't a special.  Is that a bad GUI decision?  Maybe, I don't know.  But strictly speaking, if OP is talking about the attack dialog, it's not a bug.

**Comment by rmj555** [comment]
When I put the cursor over the unit, the unit description on the right does not say the weapon is arcane.  It does so in 1.14.

**Comment by CelticMinstrel** [comment]
> A paladin also has a lance though. Arguably, holy water should be usable to make that weapon arcane too.

Ah, does it? In that case, I guess the correct filter is something like

```
[has_weapon]
    range=melee
    [not]
        type=arcane
    [/not]
[/has_weapon]
```

**Comment by CelticMinstrel** [comment]
> Weapons aren't described as 'arcane' or 'pierce' in the attack dialog. Specials get a string, but 'arcane' isn't a special. Is that a bad GUI decision? Maybe, I don't know. But strictly speaking, if OP is talking about the attack dialog, it's not a bug.

Pretty sure that info is somewhere in the attack dialog?

> When I put the cursor over the unit, the unit description on the right does not say the weapon is arcane. It does so in 1.14.

It might be conveyed by an icon now, rather than text.

**Comment by newfrenchy83** [comment]
i tried to pick holy water with paladins in htttt with version 1.15.9dev+ and it is worl perfectly now, only the unit who don't have melle attack of type diffrent of arcane are rejected. A bug in engine reolved perhaps?

**Comment by sevu** [comment]
That code changed last 5 years ago:https://github.com/wesnoth/wesnoth/commit/09e124aa8e0b0fed5bfea02cb66a3ec0157a7c6e

@newfrenchy83
A bit farther down in is another filter, which gives the bonus only to melee. So what you described makes sense and is not a bug.

So, the current filter for the event (the linked one) is allowing to take the item always when it has an effect. If anything, I would go a step farther and _disallow_ paladins to take the hold water. It's a limited item better given to other units. Other reason would be that you're more likely to play safe with the precious paladins in that scenario if they don't have arcane on charge, and as the holy water macro is not only limited to this campaign: generally (but not here) units are better off with having different damage types available. And lastly, the L3 paladins don't need a buff to begin with, at least not an offensive one. Though there might be cases where only the mounted paladin can reach the holy water to begin with.

In short, the linked code would better look like the one posted above by celmin.
Similarly, the trident filter out units with magical+arcane+ranged - that would be the mermaid priest who is pretty close with 6-4 vs 14-2. But I'm also fine with leaving it as is, allowing players to make bad decisions for the sake of a few nice situations.

**Comment by newfrenchy83** [comment]
[has_weapon] don't work in item, when i replace has_attack] by [has_weapon] and i it ry to pick item with a Lich, this unit is not rejected, then what it is rejected with has_attack. The actual code must be understand like, if the unit has an attack melle but not arcane, she can pickup item, same if she have another melee attac k of arcane type. But the issue for 1.15.8 suggest what the wml syntax was undesrtand like [not][has_attack]range=melee[and]type=arcane. i think what the engine c++ code bugs was  resolved in 1.15.9dev+ but i don't know where.

**Comment by Wedge009** [comment]
Just did a quick test:
1. Start new HttT campaign
2. Debug-skip to S9 Valley of Death
3. Debug-create unit and pick-up holy water.

1.14.17:
* Paladin is permitted to pick up, lance attack becomes arcane (sword is already arcane)
* Swordsman is permitted to pick up, sword attack becomes arcane

1.15.17+dev:
* Paladin is permitted to pick up, lance attack becomes arcane (sword is already arcane)
* Swordsman is permitted to pick up, sword attack becomes arcane

Confirmed arcane attack in attack dialogue as well.

It's worth noting that 1.15 change the attack type in the side-bar from text to icon, so it may be harder to tell if there's a change, but it's easy to check the tool-tip to confirm.
